src/main/java/net/dries007/tfc/common/blocks/plant/FloatingWaterPlantBlock.java

```diff
@@ -38,12 +51,26 @@ protected FloatingWaterPlantBlock(Properties properties)
     @Override
     public boolean canSurvive(BlockState state, IWorldReader worldIn, BlockPos pos)
     {
-        return worldIn.getBlockState(pos.below()).getMaterial() == Material.WATER;
+        BlockState belowState = worldIn.getBlockState(pos.below());
+        return (belowState.getFluidState() != Fluids.EMPTY.defaultFluidState() && belowState.getFluidState().getType() == getFavoriteFluid());
+    }
+
+    @Override
+    @SuppressWarnings("deprecation")
+    public void entityInside(BlockState state, World worldIn, BlockPos pos, Entity entityIn)
+    {
+        super.entityInside(state, worldIn, pos, entityIn); // lifted from LilyPadBlock
+        if (worldIn instanceof ServerWorld && entityIn instanceof BoatEntity)
+        {
+            worldIn.destroyBlock(new BlockPos(pos), true, entityIn);
+        }
     }
 
     @Override
     public VoxelShape getShape(BlockState state, IBlockReader worldIn, BlockPos pos, ISelectionContext context)
     {
         return SHAPE;
     }
+
+    public abstract Fluid getFavoriteFluid();
```

Comment:
1. Change this to a predicate method on a given fluid (i.e. `isValidFluid(FluidState fluid)`)
2. In the implementation of this method, check the saved fluid against the passed in fluid.

I also don't see the case why this has to be a abstract method implemented by anon. subclass (i.e. `getPlant()` exists for a very specific call path which happens during the `Block` super constructor). Is there a reason this can't just be a property on the class?